src/main/java/com/example/bookstore/user/UserService.java

```diff
@@ -22,19 +22,64 @@ public ShoppingCart getShoppingCart(String username) {
-        cart.getItems().put(bookRepository.findById(bookISBN).orElseThrow(() -> new RuntimeException("book not found")), 1);
+        cart.getItems().merge(bookRepository.findById(bookISBN).orElseThrow(() -> new RuntimeException("book not found")), quantity, Integer::sum);
+        shoppingCartRepository.save(cart);
+    }
+
+    public void removeFromCart(String username, Integer bookISBN) {
+        ShoppingCart cart = getShoppingCart(username);
+        cart.getItems().remove(bookRepository.findById(bookISBN).orElseThrow(() -> new RuntimeException("book not found")));
         shoppingCartRepository.save(cart);
     }
 
     public void checkout(String username) {
         ShoppingCart cart = getShoppingCart(username);
-        for (Map.Entry<Book, Integer> entry : cart.getItems().entrySet()) {
-            entry.getKey().setStock(entry.getKey().getStock() - entry.getValue());
-            bookRepository.save(entry.getKey());
+        User user = this.userRepository.findUserByUsername(username).orElseThrow(() -> new RuntimeException("user not found"));
+        for (Map.Entry<Book, Integer> cartItem : cart.getItems().entrySet()) {
+
+            if (cartItem.getValue() <= cartItem.getKey().getStock()) {
+                cartItem.getKey().setStock(cartItem.getKey().getStock() - cartItem.getValue());
+            } else {
+                //if amount in cart is greater than how much is in stock, buy all available
+                cartItem.getKey().setStock(0);
+            }
+            //add this book to users list of purchased books
+            user.getPurchasedBooks().add(cartItem.getKey());
+            bookRepository.save(cartItem.getKey());
         }
         cart.getItems().clear();
         shoppingCartRepository.save(cart);
     }
-}
+
+    public Iterable<Book> getRecommendations(String username) {
+
+        List<Book> recommendedBooks = new ArrayList<>();
+        User currentUser = this.userRepository.findUserByUsername(username).orElseThrow(() -> new RuntimeException("user not found")); // Retrieve the current user
+        Set<Book> currentUserPurchasedBooks = new HashSet<>(currentUser.getPurchasedBooks()); // Get purchased books of current user
+
+        for (User user : this.userRepository.findAll()) {
+            if (user.equals(currentUser)) {
+                continue; // Skip comparing the current user to itself
+            }
+
+            Set<Book> userPurchasedBooks = new HashSet<>(user.getPurchasedBooks()); // Get purchased books of the compared user
+            // Calculate Jaccard distance
+            Set<Book> union = new HashSet<>(currentUserPurchasedBooks);
+            union.addAll(userPurchasedBooks);
+            Set<Book> intersection = new HashSet<>(currentUserPurchasedBooks);
+            intersection.retainAll(userPurchasedBooks);
+            double jaccardDistance = (double) intersection.size() / (double) union.size();
+
+            if (jaccardDistance > 0.5) {

```

Comment:
Why did you choose 0.5?

Reply:
I used it randomly, but if the purchases are similar by about 50% then we consider it.